### DISCUSSION TITLE
MQTT: How to publish to all clients?

### DISCUSSION THREAD

**Discussion by abensoft-williamthome** [zotonic/zotonic#3032] [Q&A] [answered]
Hi!

I successfully subscribe to a topic via template as `{% wire type={mqtt topic="my/topic"} action={growl text="hello"} %}` or js
```javascript
cotonic.broker.subscribe("my/topic", () => {
    console.log("hello")
})
```
In the server
```erlang
{ok, ClientId} = z_context:client_id(Context),
z_mqtt:publish([<<"bridge">>, ClientId, <<"my">>, <<"topic">>], #{}, Context).

%% or

z_mqtt:publish([<<"~client">>, <<"my">>, <<"topic">>], #{}, Context).
```
... the current client receives the information on the page, but only the client who fired the event/topic.

How to propagate the event to all clients on the current page? How to, via server, all clientes/pages subscribed in the topic `{% wire type={mqtt topic="my/topic"} action={some_action} %}` performs some action?
In 0.x we used the `~pagesession` namespace, but in 1.x we could not find a way to do this.

Thanks in advance!

**Comment by abensoft-alvaro** [comment]
Hello, 

Complementing the question, we are removing the pubzub library and the same problem occurs when we tried this via .js doing `cotonic.broker.subscribe("bridge/origin/server/topic", callback);`
and on the server side `z_mqtt:publish("server/topic", data)`.
The subscription is received by the server, but the publish event never arrives to the client. 

Our scenario is:  a modification generated by one user on a page/URL has to force a refresh to all other users accessing the same page/URL. All users are authenticated.

Any help is appreciated.

**Comment by mworrell** [comment, marked as answer]
If you to publish to all clients, then all those clients should be subscribed to a topic on the server.

```javascript
cotonic.broker.subscribe("bridge/origin/my/topic", () => {
    console.log("hello")
})
```

And then on the server publish to `my/topic`

Note that all subscribes and publishes are ACL checked, so make certain that your topic can be subscribed to by the client.

For testing you can use the `test` or `public` topics, those are publicly available.

```javascript
cotonic.broker.subscribe("bridge/origin/test/foo/bar", () => {
    console.log("hello")
})
```

**Reply by abensoft-williamthome** [reply]
Thanks, @mworrell.
`test` or `public` topics work, but I accomplished this using the model topic convention by looking at Cotonic docs and `z_mqtt_sessions_runtime:is_allowed/3`. Just `bridge/origin/my/topic` does not work.
My topic looks like so now `bridge/origin/model/mymodel/event/my/topic`.
Maybe I'm making some mistake, but the [Zotonic docs ](https://zotonic.com/en/latest/ref/modules/mod_mqtt.html) give me this example:
```javascript
cotonic.broker.subscribe("bridge/origin/foo/#", function(msg, bindings, options) { console.log(msg); });
cotonic.broker.publish("bridge/origin/foo/bar", "hello world");
```
but I could not make this work as I expected. I think there is some missing explanation about MQTT topics.

**UPDATE**

Code in my erlang site file:
```erlang
observe_acl_is_allowed(
    #acl_is_allowed{object = #acl_mqtt{topic = [
        <<"model">>, <<"mymodel">>, <<"event">>, <<"my">>, <<"topic">>
    ]}},
    _Context
) ->
    true;
observe_acl_is_allowed(#acl_is_allowed{}, _Context) ->
    undefined.
```

**Comment by mworrell** [comment]
I changed the `foo/bar` topic in the docs to `test/foo`.

For the model topics, you need to have `use` permission on the module that "owns" the model to subscribe.

From `z_mqtt_sessions_runtime.erl`:
```erlang
% Model events can be view iff the user has use permission on the module
is_allowed(subscribe, [ <<"model">>, Model, <<"event">> | _ ], Context) ->
    try
        case z_module_indexer:find(model, Model, Context) of
            {ok, #module_index{ module = zotonic_core }} ->
                true;
            {ok, #module_index{ module = Module }} ->
                z_acl:is_allowed(use, Module, Context);
            {error, enoent} ->
                false
        end
    catch
        _:_ ->
            false
    end;
```

**Comment by abensoft-alvaro** [comment]
Hello Marc,

That was a question we had yesterday. This model topic refers to the Zotonic model, Cotonic model or is it a new concept? 
In the scenario where we have only a button event that should propagate to a set of users and is not related to a specific zotonic model, should we use the module name? As a suggestion, should an "event" topic exist alongside public, test, model, etc? It could call only the is_acl_allowed verification and not the acl use...

Cheers,

**Comment by mworrell** [comment]
The model is a Zotonic concept that is also used by Cotonic.

In your case, as the topic is not associated with a particular model, I would make a special topic.

It is good to make a list of topics that should be interesting or should have some default access control attached to them.

We have already the `model`, `test`, `public` and `user`.  Top-level topics. Maybe we could add some more convenient topics to these.  Especially as adding ACL checks to the code is quite some work (and easy to make errors).
